### ISSUE TITLE
models.StandardTimingModel problem?

### ISSUE THREAD

**Issue by xuhengdada** [nanograv/PINT#320]
**When I was doing:**
m = models.StandardTimingModel
m.read_parfile("NGC6440E.par")
m.phase(tt)

**and I got a problem:**

---------------------------------------------------------------------------
UnitsError                                Traceback (most recent call last)
<ipython-input-63-9bb64dd30ad9> in <module>()
----> 1 m.phase(tt)

/usr/local/lib/python2.7/site-packages/pint-0.5.2+181.g6b8fbcd-py2.7-macosx-10.12-x86_64.egg/pint/models/timing_model.pyc in phase(self, toas)
    456         """Return the model-predicted pulse phase for the given TOAs."""
    457         # First compute the delays to "pulsar time"
--> 458         delay = self.delay(toas)
    459         phase = Phase(np.zeros(len(toas)), np.zeros(len(toas)))
    460         # Then compute the relevant pulse phases

/usr/local/lib/python2.7/site-packages/pint-0.5.2+181.g6b8fbcd-py2.7-macosx-10.12-x86_64.egg/pint/models/timing_model.pyc in delay(self, toas, cutoff_component, include_last)
    450 
    451         for df in self.delay_funcs[0:idx]:
--> 452             delay += df(toas, delay)
    453         return delay
    454 

/usr/local/lib/python2.7/site-packages/astropy/units/quantity.pyc in __array_ufunc__(self, function, method, *inputs, **kwargs)
    619         # consistent units between two inputs (e.g., in np.add) --
    620         # and the unit of the result (or tuple of units for nout > 1).
--> 621         converters, unit = converters_and_unit(function, method, *inputs)
    622 
    623         out = kwargs.get('out', None)

/usr/local/lib/python2.7/site-packages/astropy/units/quantity_helper.pyc in converters_and_unit(function, method, *args)
    481                                      "argument is not a quantity (unless the "
    482                                      "latter is all zero/infinity/nan)"
--> 483                                      .format(function.__name__))
    484             except TypeError:
    485                 # _can_have_arbitrary_unit failed: arg could not be compared

UnitsError: Can only apply 'add' function to dimensionless quantities when other argument is not a quantity (unless the latter is all zero/infinity/nan)

**same problem in bash enviroment:**
pintempo --plot psr1.par.txt psr1.tim.txt
wrong:
WARNING: failed to download http://maia.usno.navy.mil/ser7/finals2000A.all, using local IERS-B: <urlopen error timed out> [astropy.utils.iers.iers]
INFO: Adding columns ssb_obs_pos ssb_obs_vel obs_sun_pos [pint.toa]
Traceback (most recent call last):
  File "/usr/local/bin/pintempo", line 11, in <module>
    load_entry_point('pint==0.5.2+181.g6b8fbcd', 'console_scripts', 'pintempo')()
  File "/usr/local/lib/python2.7/site-packages/pint-0.5.2+181.g6b8fbcd-py2.7-macosx-10.12-x86_64.egg/pint/scripts/pintempo.py", line 38, in main
    prefit_resids = pint.residuals.resids(t, m).time_resids
  File "/usr/local/lib/python2.7/site-packages/pint-0.5.2+181.g6b8fbcd-py2.7-macosx-10.12-x86_64.egg/pint/residuals.py", line 12, in __init__
    self.phase_resids = self.calc_phase_resids(weighted_mean=weighted_mean)
  File "/usr/local/lib/python2.7/site-packages/pint-0.5.2+181.g6b8fbcd-py2.7-macosx-10.12-x86_64.egg/pint/residuals.py", line 23, in calc_phase_resids
    rs = self.model.phase(self.toas.table)
  File "/usr/local/lib/python2.7/site-packages/pint-0.5.2+181.g6b8fbcd-py2.7-macosx-10.12-x86_64.egg/pint/models/timing_model.py", line 458, in phase
    delay = self.delay(toas)
  File "/usr/local/lib/python2.7/site-packages/pint-0.5.2+181.g6b8fbcd-py2.7-macosx-10.12-x86_64.egg/pint/models/timing_model.py", line 452, in delay
    delay += df(toas, delay)
  File "/usr/local/lib/python2.7/site-packages/astropy/units/quantity.py", line 621, in __array_ufunc__
    converters, unit = converters_and_unit(function, method, *inputs)
  File "/usr/local/lib/python2.7/site-packages/astropy/units/quantity_helper.py", line 483, in converters_and_unit
    .format(function.__name__))
astropy.units.core.UnitsError: Can only apply 'add' function to dimensionless quantities when other argument is not a quantity (unless the latter is all zero/infinity/nan)
———————————————————————————————————————————————
the psr1.par.txt and psr1.tim.txt is the example data of tempo2 main page.

***The last***, why this code will giving warning in the command line pintempo:  
WARNING: failed to download http://maia.usno.navy.mil/ser7/finals2000A.all, using local IERS-B: <urlopen error timed out> [astropy.utils.iers.iers]
while in python environment do not, notice that I have download finals2000A.all and put it in /usr/local/lib/python2.7/site-packages/astropy/utils/iers/data  file.

Thank you!

**Comment by luojing1211** [comment]
How did you create your tt object.

**Comment by xuhenghb** [comment]
Follow the  example:   PINT/examples/PINT_walkthrough.ipynb

**Comment by luojing1211** [comment]
I can not reproduce the error. 
This error message is coming from the astropy unit module. Somehow your PINT is adding one variable has no unit and the other with unit. 

What is your astropy verison?

**Comment by scottransom** [comment]
I'm seeing this now too doing "python fit_NGC6440E.py".  I just upgraded to astropy 2.0.  I'm wondering if that is the issue.  I'll investigate.

**Comment by scottransom** [comment]
The issue seems to be that some delay functions (like Dispersion) are returning arrays with units (i.e. seconds) and others (i.e. Astrometry and SolarSystem_Shaipro) are returning unit-less arrays.  These should all be consistent (and I think they should probably all be with seconds!)

**Comment by luojing1211** [comment]
I did not see this before. Maybe astropy 2.0 is starting picking about this. I will make the change.

**Comment by scottransom** [comment]
There is even a comment in astrometry.py in get_d_delay_quantities() that says:
```
        # TODO: Should delay not have units of u.second?
        delay = self.delay(toas)
```
I think all of our model functions/methods should return unit-ized values.  That is one of the things we have been advertising!

**Comment by scottransom** [comment]
I think this means that all the derivative functions should return quantities (i.e. with units) as well!

**Comment by luojing1211** [comment]
The actually do have units already.

**Comment by scottransom** [comment]
OK, good.  So maybe the issue is just the delay functions

**Comment by xuhenghb** [comment]
Yeah, my astropy version is also 2.0

**Comment by luojing1211** [comment]
PR #322 is for this issue.

**Comment by luojing1211** [comment]
I think this issue should be fixed by PR #322. @scottransom  and @xuhenghb, you should try it out.

**Comment by aarchiba** [comment]
Is there a set of test cases that check the presence of suitable units on all model functions/methods? This could probably be written as a parameterized test fixture, and would give confidence that this bug can be closed.
